libs/hamilton/hamilton.cpp

```diff
@@ -1,8 +1,69 @@
-#include "hamilton.hpp"
+#include "include/hamilton.hpp"
+#include "include/atsp_solver.hpp"
 
 namespace hamilton
 {
-int findCycle() {
-    return 1;
+std::vector<std::vector<std::size_t>> findKHamiltonianExtension(std::size_t k, core::Multigraph multiGraph) {
+    ATSPSolver::Matrix multiGraphAM = multiGraph.getAdjacencyMatrix();
+    ATSPSolver::Matrix cost(multiGraphAM);
+    for (int i = 0; i < cost.size(); ++i) {
+        for (int j = 0; j < cost[0].size(); ++j) {
+            if (i == j) continue;
+            if (multiGraphAM[i][j] == 0) {
+                cost[i][j] = k;
+            } else if (k >= multiGraphAM[i][j]) {
+                cost[i][j] = k - multiGraphAM[i][j];
+            } else {
+                cost[i][j] = 0;
+            }
+
+            // ATSP algorithm can't handle 0 costs (so everything is +1)
+            cost[i][j] += 1;
+        }
+    }
+
+    std::cout << "Cost matrix:" << std::endl;
+    for (const auto& row : cost) {
+        for (size_t value : row) {
+            std::cout << value << " ";
+        }
+        std::cout << std::endl;
+    }
+
+    ATSPSolver::Matrix cycleMatrix;
+
+    try {
+        ATSPSolver solver(cost);
+        cycleMatrix = solver.solve();
+
+        std::cout << "Minimal k-hamiltonian cycle:" << std::endl;
+        for (const auto& row : cycleMatrix) {
+            for (size_t value : row) {
+                std::cout << value << " ";
+            }
+            std::cout << std::endl;
+        }
+
+        for (int i = 0; i < cost.size(); ++i) {
+            for (int j = 0; j < cost[0].size(); ++j) {
+                if (cycleMatrix[i][j] != 0) {
+                    // ATSP algorithm can't handle 0 costs
+                    cycleMatrix[i][j] -= 1;
+                }
+            }
+        }
```

Comment:
Consider changing this part to
```c++
for (int i = 0; i < cost.size(); ++i) {
    for (int j = 0; j < cost[0].size(); ++j) {
        if (cycleMatrix[i][j] > 0) {
            std::cout << multiGraphAM[i][j] + cycleMatrix[i][j] - 1 << " ";
        } else {
            std::cout << 0 << " ";
        }            
    }
    std::cout << std::endl;
}
```
to make the output clearer.

Reply:
You are right, I will have it in mind. I plan bigger changes in hamilton lib design in next PR I will take care of that then.